nexus/social_media/services.py

```diff
@@ -0,0 +1,173 @@
+# Third Party Stuff
+import requests
+from django.conf import settings
+from django.utils import timezone
+from rest_framework import status
+
+# nexus Stuff
+from nexus.base import exceptions as exc
+from nexus.social_media.models import Post
+
+
+def upload_image_to_linkedin(author, headers, post, api_url_base, linkedin):
+    """Upload image to linkedin for given post.
+
+    :param author: Owner of the post to be used for response POST data.
+
+    :param headers: HTTP headers required for successful HTTP request.
+
+    :param post: Post model instance.
+
+    :param api_url_base: Base url for LinkedIn API.
+
+    :param linkedin: LINKEDIN_AUTH object taken from settings.
+
+    :returns: A valid HTTP response.
+
+    """
+    post_data_assets = {
+        "registerUploadRequest": {
+            "recipes": [
+                "urn:li:digitalmediaRecipe:feedshare-image"
+            ],
+            "owner": author,
+            "serviceRelationships": [
+                {
+                    "relationshipType": "OWNER",
+                    "identifier": "urn:li:userGeneratedContent"
+                }
+            ]
+        }
+    }
+
+    api_url_assets = f"{api_url_base}assets?action=registerUpload"
+
+    response_assets = requests.post(api_url_assets, json=post_data_assets,
+                                    headers=headers)
+    if response_assets.status_code == status.HTTP_200_OK:
+        json_response_assets = response_assets.json()
+        uploadUrl = json_response_assets.get('value')\
+                                        .get('uploadMechanism')\
+                                        .get('com.linkedin.digitalmedia.uploading.MediaUploadHttpRequest')\
+                                        .get('uploadUrl')
+
+        asset = json_response_assets.get('value').get('asset')
+        response_upload_image = requests.post(uploadUrl, data=post.image.file.read(),
+                                              headers={'Authorization': f"Bearer {linkedin['access_token']}"})
+        return (response_upload_image, asset)
```

Comment:
Is it necessary to return 2 things from this function?

Reply:
Yes, we need two things from this function. Because `response_upload_image` and `asset` both are necessary to successfully upload the image to LinkedIn if all the api hits to LinkedIn API is going well.

Reply:
I would try to break this into two functions if necessary.

Reply:
What does `asset` refers to here exactly?

Reply:
When doing an Image share on LinkedIn, we need to go through following steps
1. Registering the image to be uploaded
 To perform this, we need to send a `POST` request to the `assets` API. The endpoint for the same is `https://api.linkedin.com/v2/assets?action=registerUpload`.
A valid response contain the `uploadUrl` and `asset` key which is required to perform next steps.
2. Upload Image to LinkedIn
We need to send `POST` request to the `uploadUrl` fetched from step 1 with the binary image data to upload the image.
3. Once we get a successful response from step 2, we have to use the `asset` fetched from step 1 to attach the image to our share by sending `POST` request to our main API. The endpoint for the same is `https://api.linkedin.com/v2/ugcPosts` 
Hence if we are including the image in our share, we need to add a `media` key which includes the value of `asset`, in our request body, otherwise there is no need of `media` key and `asset` (for only text share)

Reference:- https://docs.microsoft.com/en-us/linkedin/consumer/integrations/self-serve/share-on-linkedin